parser_library/src/processing/statement_analyzers/hit_count_analyzer.cpp

```diff
@@ -0,0 +1,192 @@
+{
+    auto res_stmt = statement.access_resolved();
+    if (!res_stmt)
+        return m_stmt_type;
+
+    const auto macro_init_incrementer = [&nest_level = m_macro_nest_level](auto resolved_stmt) {
+        if (resolved_stmt->opcode_ref().value != context::id_storage::well_known::MACRO)
+            return false;
+
+        ++nest_level;
+        return true;
+    };
+
+    static constexpr auto range_adj = [](range r) {
+        r.end.column = 72;
+        return r;
+    };
+
+    switch (m_stmt_type)
+    {
+        case hlasm_plugin::parser_library::processing::hit_count_analyzer::statement_type::REGULAR:
+            if (macro_init_incrementer(res_stmt))
+            {
+                m_stmt_type = statement_type::MACRO_INIT;
+                m_last_macro_init_r = res_stmt->stmt_range_ref();
+            }
+            break;
+
+        case hlasm_plugin::parser_library::processing::hit_count_analyzer::statement_type::MACRO_INIT: {
+            m_stmt_type = statement_type::MACRO_NAME;
+
+            assert(m_ctx.processing_stack().frame().resource_loc);
+
+            m_macro_header_definitions.try_emplace(res_stmt->opcode_ref().value.to_string(),
+                macro_header_definition_details { *m_ctx.processing_stack().frame().resource_loc,
+                    range_adj(std::move(m_last_macro_init_r)),
+                    range_adj(res_stmt->stmt_range_ref()) });
+            break;
+        }
+
+        case hlasm_plugin::parser_library::processing::hit_count_analyzer::statement_type::MACRO_NAME:
+            m_stmt_type = statement_type::MACRO_BODY;
+            macro_init_incrementer(res_stmt);
+            break;
+
+        case hlasm_plugin::parser_library::processing::hit_count_analyzer::statement_type::MACRO_BODY:
+            if (res_stmt->opcode_ref().value == context::id_storage::well_known::MEND)
+            {
+                if (--m_macro_nest_level == 0)
+                    return std::exchange(m_stmt_type, statement_type::REGULAR);
+            }
+            else
+                macro_init_incrementer(res_stmt);
+            break;
+
+        default:
+            break;
+    }
+
+    return m_stmt_type;
+}
+
+void hit_count_analyzer::analyze(
+    const context::hlasm_statement& statement, statement_provider_kind prov_kind, processing_kind proc_kind, bool)
+{
+    bool can_be_external_macro = false;
+
+    switch (get_stmt_type(statement))
+    {
+        using enum processing::hit_count_analyzer::statement_type;
+        case REGULAR:
+            break;
+
+        case MACRO_BODY:
+            can_be_external_macro = true;
+            break;
+
+        default:
+            return;
+    }
+
+    auto stmt_range = get_range(&statement, prov_kind, proc_kind, m_ctx);
+    if (!stmt_range)
+        return;
+
+    if (auto macro_invo = m_ctx.this_macro(); macro_invo)
+        emplace_macro_header_definitions(macro_invo->id);
```

Comment:
This will be execute for every single macro statement, right?

Reply:
It is executed for every "regular" statement -> every statement besides macro definition statements.

If we're inside a macro then the `m_macro_header_definitions` is searched in order to see if the "header" definition for this macro exists. If it does, then the current macro entry is removed from `m_macro_header_definitions` so it cannot be found and processed again.

Reply:
I was actually thinking about adding a more light-weight function returning id_index with the current macro (let's say current_macro_id() const), empty id_index if none is currently executing.

Reply:
Alright, added `current_macro_id()` and renamed `this_macro()` to `current_macro()` for consistency